celery task to update sailthru purchase record

uzairr opened the pull request:
## [LEARNER-2694](https://openedx.atlassian.net/browse/LEARNER-2694)

### Description
Celery task to update sailthru purchase record

After a recent change in enrollment process, edX is only keeping sailthru record for the order against which a basket is created and it is not creating a basket against audit enrollment so it is no longer keeping its audit enrollment record on sailthru.To keep updated sailthru a celery task is created that will update sailthru user profile in case of enroll/un-enroll of any audit/honor enrollment course.

### Sandbox
https://celery-task.sandbox.edx.org

### Testing
1- Go to [Sandbox](https://celery-task.sandbox.edx.org)
2- Enable switch 'sailthru.audit_purchase_enabled' using [admin](https://celery-task.sandbox.edx.org/admin/waffle/switch/)
3- Enroll in the course and check [Sailthru](https://my.sailthru.com/reports/user_lookup?id=uzairr%40yahoo.com)
4- One should see the purchase record in the purchase history
5- if the switch from the [admin](https://celery-task.sandbox.edx.org/admin/waffle/switch/) is disabled, one should not see the purchase record on [Sailthru](https://my.sailthru.com/reports/user_lookup?id=uzairr%40yahoo.com)

###Admin Credentials:
username: staff
password: edx

###User Credentials:
username: uzairr@yahoo.com
password: edx

Note: Be sure to check the user-record on edx-Dev env on Sailthru.
- [x] Unit

uzairr commented:
jenkins run bokchoy

schenedx commented on common/djangoapps/student/cache.py:
```diff
@@ -0,0 +1,62 @@
+        try:
+            self[key] = CacheObject(value, duration)
+        finally:
+            lock.release()
```
#nit new line

schenedx commented on common/djangoapps/student/tasks.py:
```diff
@@ -48,3 +58,228 @@ def send_activation_email(self, subject, message, from_address, dest_addr):
+    if mode == 'audit' or mode == 'honor':
+        # free enroll
+        new_enroll = True
+        send_template = config.sailthru_enroll_template
```
I believe the code in ecom-worker https://github.com/edx/ecommerce-worker/blob/master/ecommerce_worker/sailthru/v1/tasks.py#L263 test out condition more than these. Why only test out `audit` or `honor`?

uzairr replied:
Now, it make sense to remove this condition https://github.com/edx/ecommerce-worker/blob/master/ecommerce_worker/sailthru/v1/tasks.py#L267 as now, there is no need to check whether, mode is audit or honor

schenedx commented on common/djangoapps/student/tests/test_tasks.py:
```diff
@@ -51,3 +55,87 @@ def test_send_email(self, mock_log):
+        return self.code
+
+
+class SailthruTests(TestCase):
```
Can you add a test case if the enrollment mode is `verified`, we **do not** call Sailthru api purchase API?

schenedx commented on common/djangoapps/student/tests/test_tasks.py:
```diff
@@ -51,3 +55,97 @@ def test_send_email(self, mock_log):
+
+    @mock.patch('sailthru.sailthru_client.SailthruClient.purchase')
+    def test_verify_dependency_on_flag(self, mock_sailthru_purchase):
+        update_sailthru(None, None, self.user, 'verified', self.course_id)
```
Theses should be separate into two different unit tests.

schenedx commented on common/djangoapps/student/tasks.py:
```diff
@@ -48,3 +62,224 @@ def send_activation_email(self, subject, message, from_address, dest_addr):
         raise Exception
+
+
+@receiver(ENROLL_STATUS_CHANGE)
```
I don't believe this belongs to the task.py. This receiver belongs to the signal.py

uzairr commented:
jenkins run a11y

uzairr commented:
jenkins run all

schenedx commented on common/djangoapps/student/tests/test_tasks.py:
```diff
@@ -1,12 +1,10 @@
 from boto.exception import NoAuthHandlerFound
 from django.conf import settings
 from django.test import TestCase
-
```
Please revert your changes on this file.

schenedx commented on lms/djangoapps/email_marketing/tests/test_signals.py:
```diff
@@ -523,3 +525,101 @@ def test_modify_email(self, mock_update_user):
+            i: waffle switch is 'False'
+            ii: waffle switch is 'True' and course_mode is 'verified'
+        """
+        update_sailthru(None, None, self.user, 'verified', self.course_id)
```
I do feel like this line and next belong to a different test case. Can we separate these two lines out into its own test case?

schenedx commented on lms/djangoapps/email_marketing/signals.py:
```diff
@@ -25,6 +27,28 @@
+    Returns:
+        None
+    """
+    if WAFFLE_SWITCHES.is_enabled(SAILTHRU_AUDIT_PURCHASE_ENABLED) and mode != 'verified':
```
Checking on mode==`verified` is not enough.
You should check all modes with price. 
Maybe it's easier make the condition as `mode in ('audit', 'honor')`. By the way, please don't use strings. Use the constants defined at https://github.com/edx/edx-platform/blob/master/common/djangoapps/course_modes/models.py#L159

uzairr commented:
@schenedx actually, i was looking django cache for some time and after that i came to know that the thing that i want to achieve can easily be implemented using django cache so i updated my PR.Previouly, i don't know much about django cache that's why i implemented my own.

schenedx commented on lms/djangoapps/email_marketing/signals.py:
```diff
@@ -7,24 +7,54 @@
+HONOR = 'honor'
+
+# Modes utilized for audit/free enrollments
+AUDIT_MODES = [AUDIT, HONOR]
```
Why do you still defined those modes by strings? Why not use the link to the course_mode module I gave you?

schenedx commented on lms/djangoapps/email_marketing/tests/test_signals.py:
```diff
@@ -523,3 +525,103 @@ def test_modify_email(self, mock_update_user):
+
+    @patch('openedx.core.djangoapps.waffle_utils.WaffleSwitchNamespace.is_enabled')
+    @patch('sailthru.sailthru_client.SailthruClient.purchase')
+    def test_verify_dependency_on_flag(self, mock_sailthru_purchase, flag):
```
The name of this unit test is wrong. Please update

uzairr replied:
oh my bad.
